control_panel_api/tests/test_aws_s3_access_policy.py

```diff
@@ -0,0 +1,210 @@
+        's3:ListAllMyBuckets',
+    ]
+
+    READ_ACTIONS = [
+        's3:GetObject',
+        's3:GetObjectAcl',
+        's3:GetObjectVersion',
+    ]
+
+    WRITE_ACTIONS = [
+        's3:DeleteObject',
+        's3:DeleteObjectVersion',
+        's3:PutObject',
+        's3:PutObjectAcl',
+        's3:RestoreObject',
+    ]
+
+    def assert_can_access_console(self, document):
+        for action in self.CONSOLE_ACTIONS:
+            self.assert_allowed(document, action, 'arn:aws:s3:::*')
+
+    def assert_can_list(self, document, resource):
+        self.assert_allowed(document, 's3:ListBucket', resource)
+
+    def assert_cannot_list(self, document, resource):
+        self.assert_not_allowed(document, 's3:ListBucket', resource)
+
+    def assert_can_read(self, document, resource):
+        for action in self.READ_ACTIONS:
+            self.assert_allowed(document, action, f'{resource}/*')
+
+    def assert_cannot_read(self, document, resource):
+        for action in self.READ_ACTIONS:
+            self.assert_not_allowed(document, action, f'{resource}/*')
+
+    def assert_can_write(self, document, resource):
+        for action in self.WRITE_ACTIONS:
+            self.assert_allowed(document, action, f'{resource}/*')
+
+    def assert_cannot_write(self, document, resource):
+        for action in self.WRITE_ACTIONS:
+            self.assert_not_allowed(document, action, f'{resource}/*')
+
+
+class S3AccessPolicyTestCase(S3AccessTestCase):
+
+    def test_blank_policy(self):
+        policy = S3AccessPolicy()
+        document = policy.document
+
+        self.assert_can_access_console(document)
+        self.assert_has_not_sid(document, 'list')
+        self.assert_has_not_sid(document, 'readonly')
+        self.assert_has_not_sid(document, 'readwrite')
+
+    def test_granting_access(self):
+        arns = [
+            ('arn:aws:s3:::readonly-1', False),
+            ('arn:aws:s3:::readonly-2', False),
+            ('arn:aws:s3:::readwrite-1', True),
+            ('arn:aws:s3:::readwrite-2', True),
+        ]
+
+        policy = S3AccessPolicy()
+        for (arn, readwrite) in arns:
+            policy.grant_access(arn, readwrite=readwrite)
+
+        document = policy.document
+        self.assert_can_access_console(document)
+
+        for (arn, readwrite) in arns:
+            self.assert_can_list(document, arn)
+            self.assert_can_read(document, arn)
+            if readwrite:
+                self.assert_can_write(document, arn)
+            else:
+                self.assert_cannot_write(document, arn)
+
+    def test_granting_access_again_with_different_level(self):
+        ARN = 'arn:aws:s3:::test-bucket-1'
+
+        policy = S3AccessPolicy()
+        policy.grant_access(ARN, readwrite=True)
+
+        document = policy.document
+        self.assert_can_access_console(document)
+        self.assert_can_list(document, ARN)
+        self.assert_can_read(document, ARN)
+        self.assert_can_write(document, ARN)
```

Comment:
This group of assertions seems to be repeated in a couple of places - could it be replaced with `self.assert_has_readwrite_access(document, arn)` and/or `self.assert_has_readonly_access(document, arn)`?

Reply:
I wasn't sure if I was overdoing with the assertions "library" so I didn't want yet more layers.

Having said that, I was thinking about similar so I like this idea, so I'll add these 2 methods 👍